liquibase-core/src/main/java/liquibase/snapshot/JdbcDatabaseSnapshot.java

```diff
@@ -885,35 +887,55 @@ boolean shouldBulkSelect(String schemaKey, ResultSetCache resultSetCache) {
 
                 String jdbcSchemaName = ((AbstractJdbcDatabase) database).getJdbcSchemaName(catalogAndSchema);
                 String jdbcTableName = database.escapeStringForDatabase(tableName);
-                String sql = "SELECT  NULL AS TABLE_CAT, c.OWNER AS TABLE_SCHEMA,c.OWNER, c.TABLE_NAME, c.COLUMN_NAME, NULLABLE,"
-                    + " ac.VALIDATED as VALIDATED FROM ALL_TAB_COLS c "
-                    + " JOIN ALL_COL_COMMENTS cc ON c.OWNER = cc.OWNER AND c.TABLE_NAME = cc.TABLE_NAME AND c.COLUMN_NAME = cc.COLUMN_NAME "
-                    + " LEFT JOIN all_cons_columns acc ON c.OWNER = acc.OWNER AND c.TABLE_NAME = acc.TABLE_NAME AND c.COLUMN_NAME = acc.COLUMN_NAME "
-                    + " LEFT JOIN all_constraints ac ON c.OWNER = ac.OWNER AND c.TABLE_NAME = ac.TABLE_NAME AND acc.CONSTRAINT_NAME = ac.CONSTRAINT_NAME ";
+                String sqlToSelectNotNullConstraints = "SELECT  NULL AS TABLE_CAT, atc.OWNER AS TABLE_SCHEMA, atc.OWNER, atc.TABLE_NAME, " +
+                        "atc.COLUMN_NAME, NULLABLE, ac.VALIDATED as VALIDATED, ac.SEARCH_CONDITION FROM ALL_TAB_COLS atc " +
+                        "JOIN all_cons_columns acc ON atc.OWNER = acc.OWNER AND atc.TABLE_NAME = acc.TABLE_NAME AND atc.COLUMN_NAME = acc.COLUMN_NAME " +
+                        "JOIN all_constraints ac ON atc.OWNER = ac.OWNER AND atc.TABLE_NAME = ac.TABLE_NAME AND acc.CONSTRAINT_NAME = ac.CONSTRAINT_NAME ";
 
                 if (!bulk || getAllCatalogsStringScratchData() == null) {
-                    sql += " WHERE c.OWNER='" + jdbcSchemaName + "' AND hidden_column='NO' AND CONSTRAINT_TYPE='C' "
-                        + " AND c.NULLABLE = 'Y' and search_condition is not null";
+                    sqlToSelectNotNullConstraints += " WHERE atc.OWNER='" + jdbcSchemaName + "' AND atc.hidden_column='NO' AND ac.CONSTRAINT_TYPE='C'  and ac.search_condition is not null ";
                 } else {
-                    sql += " WHERE c.OWNER IN ('" + jdbcSchemaName + "', " + getAllCatalogsStringScratchData() + ") "
-                        + " AND hidden_column='NO' AND CONSTRAINT_TYPE='C' AND c.NULLABLE = 'Y' and search_condition is not null";
+                    sqlToSelectNotNullConstraints += " WHERE c.OWNER IN ('" + jdbcSchemaName + "', " + getAllCatalogsStringScratchData() + ") "
+                        + " AND atc.hidden_column='NO' AND ac.CONSTRAINT_TYPE='C'  and ac.search_condition is not null ";
                 }
```

Comment:
I'd eliminate above `if` in favor of always using IN (sometime it will just contain single element; I believe that Oracle will optimize this case to be equivalent to `=`)

Reply:
👍